flush_writebuf() failed Broken pipe

I am running the latest master on MacOS; almost everything seems to be working for me now, launching on the master (192.168.0.1):

```
$ sudo icecc-scheduler -n frutti -u daemon -vvv
$ sudo iceccd -n frutti -m 3 -vvv -s localhost -N skree -u daemon
```

and on the slave (192.168.0.2):

```
$ sudo iceccd -n frutti -m 3 -vvv -s 192.168.0.1 -N wingy -u daemon
```

But although the slave shakes hands and even gets the environment from the master, it never seems to actually build anything.  My build command output is full of these:

```
ICECC[39986] 16:56:08: flush_writebuf() failed Broken pipe
ICECC[39986] 16:56:08: write of source chunk to host 192.168.0.2
ICECC[39986] 16:56:08: failed  Broken pipe
ICECC[39986] 16:56:08: got exception Error 15 - write to host failed (192.168.0.2) 
```

and my slave log looks like this:

```
[2909] 16:55:12: ICECREAM daemon 1.1rc2 starting up (nice level 5) 
[2909] 16:55:12: 8 CPU(s) online on this server
[2909] 16:55:12: allowing up to 10 active jobs
[2909] 16:55:12: not detaching
[2909] 16:55:12: entered process group
[2909] 16:55:12: ignoring localhost lo0
[2909] 16:55:12: broadcast en0 17.114.131.255
[2909] 16:55:12: broadcast bridge0 192.168.0.255
[2909] 16:55:12: Netnames:
[2909] 16:55:12: frutti
[2909] 16:55:12: scheduler is on 192.168.0.1:8765 (net )
[2909] 16:55:12: scheduler not yet found.
[2909] 16:55:15: scheduler not yet found.
[2909] 16:55:18: Connected to scheduler (I am known as 192.168.0.2)
[2909] 16:55:44: accepted 7 192.168.0.1
[2909] 16:55:44: handle_transfer_env
[2909] 16:55:44: pid 2910
[2909] 16:55:44: got pid 2910
[2909] 16:55:44: handle_transfer_env_done
[2909] 16:55:44: installed_size: 59830408
[2909] 16:55:44: installed Darwin16_x86_64/fc70ec139a0b6f7f67e5f0a262357823 size: 59830408 all: 59830408
[2909] 16:55:44: reannounce_environments 
[2909] 16:55:44: Verify environment done, success, environment fc70ec139a0b6f7f67e5f0a262357823 (Darwin16_x86_64)
[2909] 16:55:44: requests--1
[2909] 16:55:44: handle connection returned 2912
[2912] 23:55:44: <parent, waiting>
[2912] 23:55:44: timeout <= 0
[2912] 23:55:44: !wait_for_msg()
[2912] 23:55:44: timeout <= 0
[2912] 23:55:44: !wait_for_msg()
[2912] 23:55:44: timeout <= 0
[2912] 23:55:44: !wait_for_msg()
[2912] 23:55:44: timeout <= 0
[2912] 23:55:44: !wait_for_msg()
[2912] 23:55:44: </parent, waiting: 20ms>
[2909] 16:55:45: accepted 7 192.168.0.1
[2909] 16:55:45: requests--7
[2909] 16:55:45: handle connection returned 2914
[2914] 23:55:45: <parent, waiting>
[2914] 23:55:45: timeout <= 0
[2914] 23:55:45: !wait_for_msg()
[2914] 23:55:45: timeout <= 0
[2914] 23:55:45: !wait_for_msg()
[2914] 23:55:45: timeout <= 0
[2914] 23:55:45: !wait_for_msg()
[2914] 23:55:45: timeout <= 0
[2914] 23:55:45: !wait_for_msg()
[2914] 23:55:45: </parent, waiting: 14ms>
[2909] 16:55:45: accepted 7 192.168.0.1
[2909] 16:55:45: requests--14
[2909] 16:55:45: handle connection returned 2916
[2916] 23:55:45: <parent, waiting>
[2916] 23:55:45: timeout <= 0
[2916] 23:55:45: !wait_for_msg()
[2916] 23:55:45: timeout <= 0
[2916] 23:55:45: !wait_for_msg()
[2916] 23:55:45: timeout <= 0
[2916] 23:55:45: !wait_for_msg()
[2916] 23:55:45: timeout <= 0
[2916] 23:55:45: !wait_for_msg()
[2916] 23:55:45: </parent, waiting: 14ms>
[2909] 16:55:46: accepted 7 192.168.0.1
[2909] 16:55:46: requests--16
...etc...
```

Any help would be most appreciated!

I tried to repeat this case by inserting the line:  “this is illegal code”.  I don’t think there are any surprises here, but I figured I’d show this anyway just in case it tells you something.  Any other debugging you want me to do?


Before:
ICECC[176] 10:09:50: <send compile_file>
ICECC[176] 10:09:50: </send compile_file: 0ms>
ICECC[176] 10:09:50: <write_server_cpp from cpp>
ICECC[176] 10:09:50: </write_server_cpp from cpp: 546ms>
ICECC[176] 10:09:50: <wait for cpp>
ICECC[176] 10:09:50: </wait for cpp: 0ms>
ICECC[176] 10:09:50: <wait for cs>
ICECC[176] 10:09:53: </wait for cs: 2414ms>
ICECC[176] 10:09:53: command needs stdout/stderr workaround, recompiling locally
ICECC[176] 10:09:53: local build forced by remote exception: Error 102 - command needs stdout/stderr workaround, recompiling locally
testcase.cpp:22:1: error: expected unqualified-id before 'this'
this is illegal code
^
ERROR: ninja: build stopped: subcommand failed.


After:

ICECC[169] 10:14:42: <send compile_file>
ICECC[169] 10:14:42: </send compile_file: 0ms>
ICECC[169] 10:14:42: <write_server_cpp from cpp>
ICECC[169] 10:14:43: </write_server_cpp from cpp: 548ms>
ICECC[169] 10:14:43: <wait for cpp>
ICECC[169] 10:14:43: </wait for cpp: 0ms>
ICECC[169] 10:14:43: <wait for cs>
ICECC[169] 10:14:45: </wait for cs: 2018ms>
testcase.cpp:22:1: error: expected unqualified-id before 'this'
ICECC[169] 10:14:45: Compiled on 10.140.160.74
ERROR: ninja: build stopped: subcommand failed.



From: adam-trhon [mailto:notifications@github.com]
Sent: Sunday, March 12, 2017 10:41 AM
To: icecc/icecream <icecream@noreply.github.com>
Cc: Miller Henry <MillerHenry@JohnDeere.com>; Comment <comment@noreply.github.com>
Subject: Re: [icecc/icecream] flush_writebuf() failed Broken pipe (#167)


@dabrahams<https://github.com/dabrahams>, @sss123next<https://github.com/sss123next> can you still repeat the case? If so, please try to enable icecc logs:

export ICECC_DEBUG=info

and disable caret workaround:

export ICECC_CARET_WORKAROUND=0

in your client's terminal (where you run make). Does it print anything interesting now?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/icecc/icecream/issues/167#issuecomment-285952830>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AI8oEnL8vHopmvijmnNlqcphAqaqCSNGks5rlBIQgaJpZM4J0LKY>.

--- a/daemon/workit.cpp
+++ b/daemon/workit.cpp
@@ -382,7 +382,7 @@ int work_it(CompileJob &j, unsigned int job_stat[], MsgChannel *client, CompileR
                     if (msg->type == M_END) {
                         input_complete = true;
 
-                        if (!fcmsg) {
+                        if (!fcmsg && sock_in[1] != -1) {
                             close(sock_in[1]);
                             sock_in[1] = -1;
                         }
@@ -427,6 +427,16 @@ int work_it(CompileJob &j, unsigned int job_stat[], MsgChannel *client, CompileR
 
         int max_fd = std::max(sock_out[0], sock_err[0]);
 
+        if (sock_in[1] == -1 && fcmsg) {
+            // This state can occur when the compiler has terminated before
+            // all file input is received from the client.  The daemon must continue
+            // reading all file input from the client because the client expects it to.
+            // Deleting the file chunk message here tricks the select() below to continue
+            // listening for more file data from the client even though it is being
+            // thrown away.
+            delete fcmsg;
+            fcmsg = 0;
+        }
         if (client_fd >= 0 && !fcmsg) {
             FD_SET(client_fd, &rfds);
 
@@ -438,16 +448,27 @@ int work_it(CompileJob &j, unsigned int job_stat[], MsgChannel *client, CompileR
             // we poll it in every iteration.
         }
 
-        FD_SET(death_pipe[0], &rfds);
+        // If all file data has been received from the client then start
+        // listening on the death_pipe to know when the compiler has
+        // terminated.  The daemon can't start listening for the death of
+        // the compiler sooner or else it might close the client socket before the
+        // client had time to write all of the file data and wait for a response.
+        // The client isn't coded to properly handle the closing of the socket while
+        // sending all file data to the daemon.
+        if (input_complete) {
+            FD_SET(death_pipe[0], &rfds);
 
-        if (death_pipe[0] > max_fd) {
-            max_fd = death_pipe[0];
+            if (death_pipe[0] > max_fd) {
+                max_fd = death_pipe[0];
+            }
         }
 
         fd_set wfds, *wfdsp = 0;
         FD_ZERO(&wfds);
 
-        if (fcmsg) {
+        // Don't try to write to sock_in it if was already closed because
+        // the compile terminated before reading all of the file data.
+        if (fcmsg && sock_in[1] != -1) {
             FD_SET(sock_in[1], &wfds);
             wfdsp = &wfds;
 
@@ -501,11 +522,13 @@ int work_it(CompileJob &j, unsigned int job_stat[], MsgChannel *client, CompileR
                     }
 
                     kill(pid, SIGTERM); // Most likely crashed anyway ...
-                    return_value = EXIT_COMPILER_CRASHED;
-                    client_fd = -1;
-                    input_complete = true;
+                    if (input_complete) {
+                        return_value = EXIT_COMPILER_CRASHED;
+                    }
                     delete fcmsg;
                     fcmsg = 0;
+                    close(sock_in[1]);
+                    sock_in[1] = -1;
                     continue;
                 }
 

